File: tools/update_posts_index.py
#!/usr/bin/env python3
"""
Scan the posts/ directory, read Markdown front matter, and regenerate data/posts.json.
The output is an array of slugs sorted by date (newest first), falling back to slug.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Tuple

ROOT = Path(__file__).resolve().parent.parent
POSTS_DIR = ROOT / "posts"
OUTPUT_PATH = ROOT / "data" / "posts.json"


@dataclass
class PostMeta:
    slug: str
    date: datetime | None


def main() -> None:
    if not POSTS_DIR.exists():
        raise SystemExit(f"Posts directory not found: {POSTS_DIR}")

    posts: List[PostMeta] = []

    for path in POSTS_DIR.glob("*.md"):
        slug = path.stem
        attributes, _ = parse_front_matter(path.read_text(encoding="utf-8"))
        posts.append(PostMeta(slug=slug, date=parse_date(attributes.get("date"))))

    posts.sort(key=lambda post: (_sort_key(post.date), post.slug))
    posts.reverse()

    slugs = [post.slug for post in posts]
    OUTPUT_PATH.write_text(json.dumps(slugs, indent=2) + "\n", encoding="utf-8")
    print(f"Updated {OUTPUT_PATH} with {len(slugs)} post(s).")


def parse_front_matter(source: str) -> Tuple[Dict[str, str], str]:
    if not source.startswith("---\n"):
        return {}, source

    try:
        _, rest = source.split("---\n", 1)
        front_matter, body = rest.split("\n---", 1)
    except ValueError:
        return {}, source

    attributes: Dict[str, str] = {}
    for line in front_matter.splitlines():
        cleaned = line.strip()
        if not cleaned or cleaned.startswith("#"):
            continue
        if ":" not in cleaned:
            continue
        key, value = cleaned.split(":", 1)
        attributes[key.strip()] = value.strip()

    if body.startswith("\n"):
        body = body[1:]

    return attributes, body


def parse_date(value: str | None) -> datetime | None:
    if not value:
        return None
    try:
        return datetime.fromisoformat(value)
    except ValueError:
        try:
            return datetime.strptime(value, "%Y-%m-%d")
        except ValueError:
            return None


def _sort_key(value: datetime | None) -> float:
    return value.timestamp() if value else float("-inf")


if __name__ == "__main__":
    main()

File: tools/update-posts-index.js
#!/usr/bin/env node
/**
 * Regenerates data/posts.json by scanning Markdown files in /posts.
 * Keeps slugs sorted by front matter date (newest first), falling back to name.
 */

const fs = require("node:fs/promises");
const path = require("node:path");

const ROOT = path.resolve(__dirname, "..");
const POSTS_DIR = path.join(ROOT, "posts");
const OUTPUT_PATH = path.join(ROOT, "data", "posts.json");

async function main() {
  const entries = await fs.readdir(POSTS_DIR, { withFileTypes: true });
  const markdownFiles = entries
    .filter((entry) => entry.isFile() && entry.name.endsWith(".md"))
    .map((entry) => entry.name);

  const posts = [];

  for (const filename of markdownFiles) {
    const absolutePath = path.join(POSTS_DIR, filename);
    const raw = await fs.readFile(absolutePath, "utf8");
    const { attributes } = parseFrontMatter(raw);
    const slug = filename.replace(/\.md$/i, "");
    const parsedDate = parseDate(attributes.date);

    posts.push({
      slug,
      title: attributes.title ?? slug,
      summary: attributes.summary ?? "",
      date: attributes.date ?? "",
      category: attributes.category ?? "",
      image: attributes.image ?? attributes.cover ?? "",
      sortKey: parsedDate,
    });
  }

  posts.sort((a, b) => {
    const diff = (b.sortKey ?? 0) - (a.sortKey ?? 0);
    if (diff !== 0) {
      return diff;
    }
    return a.slug.localeCompare(b.slug);
  });

  const payload = posts.map(({ sortKey, ...rest }) => rest);
  const json = `${JSON.stringify(payload, null, 2)}\n`;

  await fs.writeFile(OUTPUT_PATH, json, "utf8");
  console.log(`Updated ${OUTPUT_PATH} with ${payload.length} post(s).`);
}

function parseFrontMatter(source) {
  const match = source.match(/^---\n([\s\S]*?)\n---\n?/);
  if (!match) {
    return { attributes: {}, body: source };
  }

  const attributes = {};
  const rawBlock = match[1];

  rawBlock.split("\n").forEach((line) => {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) {
      return;
    }
    const [key, ...rest] = trimmed.split(":");
    if (!key || rest.length === 0) {
      return;
    }
    attributes[key.trim()] = rest.join(":").trim();
  });

  const body = source.slice(match[0].length);
  return { attributes, body };
}

function parseDate(value) {
  if (!value) {
    return 0;
  }
  const timestamp = Date.parse(value);
  return Number.isFinite(timestamp) ? timestamp : 0;
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
